### PR TITLE
fix(artifacts): Handle gcs input artifact key that is a directory. Fixes #2841

### LINKED ISSUE #2841: Unable to load GCS artifact which is a not empty directory
**What happened**:

Workload which attempts to load input artifact from GCS with key that is a not empty directory fails like this:
```
Name:                test-h5rm4
Namespace:           default
ServiceAccount:      default
Status:              Error
Message:             failed to load artifacts: timed out waiting for the condition
Created:             Sun Apr 26 10:07:14 +0000 (34 seconds ago)
Started:             Sun Apr 26 10:07:14 +0000 (34 seconds ago)
Finished:            Sun Apr 26 10:07:48 +0000 (now)
Duration:            34 seconds

STEP           TEMPLATE  PODNAME     DURATION  MESSAGE
 ⚠ test-h5rm4  test      test-h5rm4  33s       failed to load artifacts: timed out waiting for the condition
```

**What you expected to happen**:

Directory artifact is recursively loaded under specified path and workload ends with a success.

**How to reproduce it (as minimally and precisely as possible)**:

Workflow template:
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: test-
spec:
  entrypoint: test
  templates:
  - name: test
    container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["hello world"]
    inputs:
      artifacts:
        - name: test-dir-artifact
          path: /test_dir/
          gcs:
            bucket: test-argo-bucket
            key: test_dir
            serviceAccountKeySecret:
              name: argo-gcs-credentials
              key: serviceAccountKey
```

1. Setup GCS bucket with `test_dir` directory that has some content.
2. Run the workflow defined above.

**Anything else we need to know?**:

This does not fail when the directory that the workflow wants to load is empty.

**Environment**:

I'm using GKE.

- Argo version:
```
$ argo version
argo: v2.8.0-rc2+4126d22.dirty
  BuildDate: 2020-04-23T22:28:06Z
  GitCommit: 4126d22b6f49e347ae1a75dd3ad6f484bee30f11
  GitTreeState: dirty
  GitTag: v2.8.0-rc2
  GoVersion: go1.13.4
  Compiler: gc
  Platform: linux/amd64
```
- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: "2020-04-16T11:56:40Z"
  compiler: gc
  gitCommit: 52c56ce7a8272c798dbc29846288d7cd9fbae032
  gitTreeState: clean
  gitVersion: v1.18.2
  goVersion: go1.13.9
  major: "1"
  minor: "18"
  platform: linux/amd64
serverVersion:
  buildDate: "2020-02-21T18:01:40Z"
  compiler: gc
  gitCommit: 145f9e21a4515947d6fb10819e5a336aff1b6959
  gitTreeState: clean
  gitVersion: v1.14.10-gke.27
  goVersion: go1.12.12b4
  major: "1"
  minor: 14+
  platform: linux/amd64
```

**Logs**

```
argo get <workflowname>
Name:                test-h5rm4
Namespace:           default
ServiceAccount:      default
Status:              Error
Message:             failed to load artifacts: timed out waiting for the condition
Created:             Sun Apr 26 10:07:14 +0000 (9 minutes ago)
Started:             Sun Apr 26 10:07:14 +0000 (9 minutes ago)
Finished:            Sun Apr 26 10:07:48 +0000 (8 minutes ago)
Duration:            34 seconds

STEP           TEMPLATE  PODNAME     DURATION  MESSAGE
 ⚠ test-h5rm4  test      test-h5rm4  33s       failed to load artifacts: timed out waiting for the condition


kubectl logs <failedpodname> -c init
time="2020-04-26T10:07:15Z" level=info msg="Starting Workflow Executor" version=v2.7.5+ede163e.dirty
time="2020-04-26T10:07:15Z" level=info msg="Creating a docker executor"
time="2020-04-26T10:07:15Z" level=info msg="Executor (version: v2.7.5+ede163e.dirty, build_date: 2020-04-21T01:12:08Z) initialized (pod: default/test-h5rm4) with template:\n{\"name\":\"test\",\"arguments\":{},\"inputs\":{\"artifacts\":[{\"name\":\"test-dir-artifact\",\"path\":\"/test_dir/\",\"gcs\":{\"bucket\":\"test-argo-bucket\",\"serviceAccountKeySecret\":{\"name\":\"argo-gcs-credentials\",\"key\":\"serviceAccountKey\"},\"key\":\"test_dir\"}}]},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"docker/whalesay:latest\",\"command\":[\"cowsay\"],\"args\":[\"hello world\"],\"resources\":{}}}"
time="2020-04-26T10:07:15Z" level=info msg="Start loading input artifacts..."
time="2020-04-26T10:07:15Z" level=info msg="Downloading artifact: test-dir-artifact"
time="2020-04-26T10:07:15Z" level=info msg="GCS Load path: /argo/inputs/artifacts/test-dir-artifact.tmp, key: test_dir"
time="2020-04-26T10:07:16Z" level=warning msg="Failed to download objects from GCS: download object: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: file exists"
time="2020-04-26T10:07:18Z" level=info msg="GCS Load path: /argo/inputs/artifacts/test-dir-artifact.tmp, key: test_dir"
time="2020-04-26T10:07:18Z" level=warning msg="Failed to download objects from GCS: download object: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: file exists"
time="2020-04-26T10:07:22Z" level=info msg="GCS Load path: /argo/inputs/artifacts/test-dir-artifact.tmp, key: test_dir"
time="2020-04-26T10:07:22Z" level=warning msg="Failed to download objects from GCS: download object: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: file exists"
time="2020-04-26T10:07:31Z" level=info msg="GCS Load path: /argo/inputs/artifacts/test-dir-artifact.tmp, key: test_dir"
time="2020-04-26T10:07:31Z" level=warning msg="Failed to download objects from GCS: download object: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: file exists"
time="2020-04-26T10:07:47Z" level=info msg="GCS Load path: /argo/inputs/artifacts/test-dir-artifact.tmp, key: test_dir"
time="2020-04-26T10:07:47Z" level=warning msg="Failed to download objects from GCS: download object: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: mkdir /argo/inputs/artifacts/test-dir-artifact.tmp/: file exists"
time="2020-04-26T10:07:47Z" level=error msg="executor error: timed out waiting for the condition"
time="2020-04-26T10:07:47Z" level=info msg="Alloc=5187 TotalAlloc=18517 Sys=70848 NumGC=6 Goroutines=6"
time="2020-04-26T10:07:47Z" level=fatal msg="timed out waiting for the condition"

kubectl logs <failedpodname> -c wait
Error from server (BadRequest): container "wait" in pod "test-h5rm4" is waiting to start: PodInitializing
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by hashkanna**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to USERS.md.

### CODE REVIEW

**Comment by hashkanna** [comment]
Fix for #2841

**Comment by whynowy** [comment]
@hashkanna - Appreciate if you could share more information about `gcs objects that end with "/"`.

File: workflow/artifacts/gcs/gcs.go [hunk 1]
```diff
@@ -144,7 +144,10 @@ func listByPrefix(client *storage.Client, bucket, prefix, delim string) ([]strin
 		if err != nil {
 			return nil, err
 		}
-		results = append(results, attrs.Name)
+		// ignore gcs objects that end with "/" as they point to directory names
+		if !strings.HasSuffix(attrs.Name, "/") {
+			results = append(results, attrs.Name)
```
Comment (alexec) [inline workflow/artifacts/gcs/gcs.go:149]: needs tests

**Comment by hashkanna** [comment]
Thanks @alexec. I am still a newbie to Go and Argo. might take me a bit of time to write the Mock tests to simulate GCS behavior. Happy to accept any help or guide someone with my GCS knowledge, if this needs to be fixed quickly. 

@whynowy: "A trailing slash always indicates a directory, which is compliant with POSIX.1-2008." - as you can see it mentioned in this link. https://beam.apache.org/releases/javadoc/2.13.0/org/apache/beam/sdk/extensions/gcp/util/gcsfs/GcsPath.html

If you do a "gsutil ls" on a bucket that has directories and files. the ls will list the directory names with a trailing slash and the files without a trailing slash. The same happens in this code with the listByPrefix func.
https://github.com/argoproj/argo/blob/master/workflow/artifacts/gcs/gcs.go#L117

Example command:
gsutil ls gs://gcp-public-data-landsat/

The problem with this is that, the GCS directory is being treated as a file in this code and a file is created in [line 117](https://github.com/argoproj/argo/blob/master/workflow/artifacts/gcs/gcs.go#L117) .
but during the next iteration in the for loop in [line 88](https://github.com/argoproj/argo/blob/master/workflow/artifacts/gcs/gcs.go#L88)  for the files in the GCS directory, a call to MkdirAll is made in [line 104](https://github.com/argoproj/argo/blob/master/workflow/artifacts/gcs/gcs.go#L104) but since the top level directory had already been created as a file in the container, a "File Exists" error occurs. In other words, you could simulate it in your terminal by touching a file and then trying to create a directory of the same name. "touch testdir; mkdir testdir"

**Comment by hadim** [comment]
I am also running into that issue. Are Docker images of cargo built against PR of this repo so I can just replace the image tag in my argo installation manifests?

**Comment by alexec** [comment]
This issue is now stale and could do with a new owner.

**Comment by agilgur5** [comment]
Replaced by #10214

**Pull request closed without merging**